File: frontend/src/app/core/types/surveillance.ts
/**
 * On-disk shape of a surveillance, mirroring the JSON files stored in the
 * per-surveillance GitHub repo. Kept as a single TypeScript namespace so the
 * storage layer and the UI agree on types.
 */

export interface StratigraphicUnit {
  number: number;
  type?: string | null;
  definition?: string | null;
  description?: string | null;
  interpretation?: string | null;
  materials?: string | null;
}

export interface Finding {
  id: string;
  name: string;
  description?: string | null;
  interpretation?: string | null;
  /** Date range of the *finding itself* — e.g. Roman = -0050 to 0200. */
  start_date?: string | null;
  end_date?: string | null;
  tags: Record<string, string>;
  geometry?: GeoJSON.Geometry | null;
  units: StratigraphicUnit[];
  /** ISO date (YYYY-MM-DD) when the archaeologist recorded the finding.
   * Links this evidenza to a giornale-di-scavo day. */
  recorded_on?: string | null;
}

export interface Photo {
  id: string;
  filename: string;
  caption?: string | null;
  bearing?: number | null;
  /** EXIF datetime of capture (camera clock). */
  taken_at?: string | null;
  /** GeoJSON Point in EPSG:4326, derived from EXIF when available. */
  location?: GeoJSON.Point | null;
  /** Path of the binary inside the surveillance repo, e.g. "photos/<id>.jpg". */
  path: string;
  /** Recorded MIME type of the binary. */
  content_type?: string | null;
  /** ISO date (YYYY-MM-DD) of the giornale day this photo belongs to. */
  recorded_on?: string | null;
  /** intervento (panoramica giornaliera) vs dettaglio (post-pulizia). */
  shot_type?: 'intervento' | 'dettaglio' | null;
  /** Optional link to a Finding when shot_type === 'dettaglio'. */
  finding_id?: string | null;
}

export type SurveillanceStatus = 'draft' | 'in-progress' | 'submitted' | 'archived';

export interface Surveillance {
  id: string;
  title: string;
  status: SurveillanceStatus;
  protocollo?: string | null;
  committente?: string | null;
  direttore_tecnico?: string | null;
  sabap?: string | null;
  comune?: string | null;
  provincia?: string | null;
  foglio_catastale?: string | null;
  particelle?: string | null;
  normativa?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  premessa?: string | null;
  metodologia?: string | null;
  risultati?: string | null;
  conclusioni?: string | null;
  /** GeoJSON Polygon — the watched area in EPSG:4326. */
  area?: GeoJSON.Polygon | null;

  findings: Finding[];
  photos: Photo[];
  days: DayLog[];
  tavole: Tavola[];

  created_at: string;
  updated_at: string;
}

/**
 * Giornale di scavo / assistenza — one entry per site day.
 *
 * The Sovrintendenza requires: dates, hourly presences, names of personnel,
 * brief description of operations and their location. Stored as
 * `daily/<YYYY-MM-DD>.json` in the per-surveillance repo.
 */
export interface Presence {
  name: string;
  /** direttore tecnico · archeologo · operatore · operaio · altro */
  role?: string | null;
  hours_start?: string | null;   // "08:00"
  hours_end?: string | null;     // "17:30"
  /** Total hours on site for the day. Optional manual override (covers breaks). */
  hours_total?: number | null;
}

/**
 * A "Tavola grafica" — a rendered map snapshot saved into the surveillance.
 * Distinct from Photo because the Sovrintendenza spec lists "Tavole grafiche"
 * as its own deliverable and the DOCX places them in separate sections.
 *
 * Stored as `tavole/<id>.<ext>` (binary, usually PNG) + `tavole/<id>.json`.
 */
export interface Tavola {
  id: string;
  /** Display filename used in the GitHub UI and DOCX caption fallback. */
  filename: string;
  caption?: string | null;
  /** ISO date YYYY-MM-DD when the snapshot was captured. */
  captured_on?: string | null;
  /**
   * Drives DOCX placement:
   *   insieme   → §2.1 Tavola d'insieme di posizionamento topografico
   *   storica   → §2 historical-orto comparison
   *   dettaglio → §7.1.X under the corresponding evidenza
   */
  kind: 'insieme' | 'dettaglio' | 'storica';
  /** Required when kind === 'dettaglio'. */
  finding_id?: string | null;
  /** Path in the surveillance repo (tavole/<id>.png). */
  path: string;
  content_type?: string | null;
  /** Map state at capture, for later reproduction. */
  map_state?: {
    center?: [number, number];
    zoom?: number;
    layers?: string[];
  } | null;
}

export interface DayLog {
  /** ISO date YYYY-MM-DD; also the storage filename. */
  date: string;
  presenze: Presence[];
  operazioni?: string | null;
  /** Local descriptor for where the day's work was concentrated. */
  localizzazione?: string | null;
  weather?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

/** Lightweight summary stored in the index repo (one file per surveillance). */
export interface SurveillanceIndexEntry {
  id: string;
  title: string;
  comune?: string | null;
  provincia?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  status: 'draft' | 'in-progress' | 'submitted' | 'archived';
  /** Bbox of the area in [west, south, east, north]. */
  bbox?: [number, number, number, number] | null;
  /** Full URL to the per-surveillance repo, e.g. https://github.com/owner/archaeo-pro-{uuid}. */
  repo_url: string;
  /** Owner/name pair for API calls. */
  repo: { owner: string; name: string };
  ohm_published: boolean;
  // Re-stated here even though the interface already has `status` above —
  // this is the index-side mirror, kept in sync by SurveillanceStore.
  created_at: string;
  updated_at: string;
}

File: backend/app/models/surveillance.py
from datetime import date, datetime
from uuid import UUID, uuid4

from geoalchemy2 import Geometry
from sqlalchemy import JSON, Date, DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID as PgUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base


class Surveillance(Base):
    """A *sorveglianza archeologica* — the root aggregate."""

    __tablename__ = "surveillances"

    id: Mapped[UUID] = mapped_column(PgUUID(as_uuid=True), primary_key=True, default=uuid4)

    # Header / metadata
    protocollo: Mapped[str | None] = mapped_column(String(64))
    title: Mapped[str] = mapped_column(String(255))
    committente: Mapped[str | None] = mapped_column(String(255))
    direttore_tecnico: Mapped[str | None] = mapped_column(String(255))
    sabap: Mapped[str | None] = mapped_column(String(255))  # competent SABAP office
    comune: Mapped[str | None] = mapped_column(String(128))
    provincia: Mapped[str | None] = mapped_column(String(64))
    foglio_catastale: Mapped[str | None] = mapped_column(String(64))
    particelle: Mapped[str | None] = mapped_column(String(255))
    normativa: Mapped[str | None] = mapped_column(Text)

    start_date: Mapped[date | None] = mapped_column(Date)
    end_date: Mapped[date | None] = mapped_column(Date)

    # The watched area — polygon in EPSG:4326.
    area: Mapped[object | None] = mapped_column(Geometry("POLYGON", srid=4326))

    # Free-form sections — the docgen renders them into the template.
    premessa: Mapped[str | None] = mapped_column(Text)
    metodologia: Mapped[str | None] = mapped_column(Text)
    risultati: Mapped[str | None] = mapped_column(Text)
    conclusioni: Mapped[str | None] = mapped_column(Text)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    findings: Mapped[list["Finding"]] = relationship(
        back_populates="surveillance", cascade="all, delete-orphan", order_by="Finding.created_at"
    )
    photos: Mapped[list["Photo"]] = relationship(
        back_populates="surveillance", cascade="all, delete-orphan", order_by="Photo.taken_at"
    )


class Finding(Base):
    """An archaeological finding — point/polygon with date range and OSM-ish tags.

    The OHM exporter consumes these directly.
    """

    __tablename__ = "findings"

    id: Mapped[UUID] = mapped_column(PgUUID(as_uuid=True), primary_key=True, default=uuid4)
    surveillance_id: Mapped[UUID] = mapped_column(
        PgUUID(as_uuid=True),
        ForeignKey("surveillances.id", ondelete="CASCADE"),
        index=True,
    )

    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str | None] = mapped_column(Text)
    interpretation: Mapped[str | None] = mapped_column(Text)

    # Temporal extents — partial ISO dates allowed (e.g. "-0500" or "1450").
    # Stored as text because we want pre-1AD and partial dates.
    start_date: Mapped[str | None] = mapped_column(String(32))
    end_date: Mapped[str | None] = mapped_column(String(32))

    # OSM-style tags, e.g. {"historic": "archaeological_site", "site_type": "settlement"}.
    tags: Mapped[dict] = mapped_column(JSON, default=dict)

    geom: Mapped[object | None] = mapped_column(Geometry("GEOMETRY", srid=4326))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    surveillance: Mapped["Surveillance"] = relationship(back_populates="findings")
    units: Mapped[list["StratigraphicUnit"]] = relationship(
        back_populates="finding", cascade="all, delete-orphan", order_by="StratigraphicUnit.number"
    )


class StratigraphicUnit(Base):
    """A US (Unità Stratigrafica) — the atom of an archaeological record."""

    __tablename__ = "stratigraphic_units"

    id: Mapped[UUID] = mapped_column(PgUUID(as_uuid=True), primary_key=True, default=uuid4)
    finding_id: Mapped[UUID] = mapped_column(
        PgUUID(as_uuid=True), ForeignKey("findings.id", ondelete="CASCADE"), index=True
    )

    number: Mapped[int] = mapped_column(Integer)
    type: Mapped[str | None] = mapped_column(String(64))  # positiva / negativa / struttura
    definition: Mapped[str | None] = mapped_column(Text)
    description: Mapped[str | None] = mapped_column(Text)
    interpretation: Mapped[str | None] = mapped_column(Text)
    materials: Mapped[str | None] = mapped_column(Text)

    finding: Mapped["Finding"] = relationship(back_populates="units")


class Photo(Base):
    __tablename__ = "photos"

    id: Mapped[UUID] = mapped_column(PgUUID(as_uuid=True), primary_key=True, default=uuid4)
    surveillance_id: Mapped[UUID] = mapped_column(
        PgUUID(as_uuid=True),
        ForeignKey("surveillances.id", ondelete="CASCADE"),
        index=True,
    )
    finding_id: Mapped[UUID | None] = mapped_column(
        PgUUID(as_uuid=True), ForeignKey("findings.id", ondelete="SET NULL"), nullable=True
    )

    filename: Mapped[str] = mapped_column(String(255))
    storage_path: Mapped[str] = mapped_column(String(512))
    caption: Mapped[str | None] = mapped_column(Text)

    taken_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    bearing: Mapped[float | None] = mapped_column()
    location: Mapped[object | None] = mapped_column(Geometry("POINT", srid=4326))

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    surveillance: Mapped["Surveillance"] = relationship(back_populates="photos")
